### ISSUE TITLE
Division by zero when calling igraph_average_path_length() with empty graph

### ISSUE THREAD

**Issue by grievejia** [igraph/igraph#1028]
Running the following program with igraph master(5a8f0c1) and UndefinedBehaviorSanitizer will result in a division by zero being reported:
```c
#include <igraph.h>

int main() {
  igraph_t g;
  igraph_small(&g, 1, 1, -1);
  igraph_real_t res;
  igraph_average_path_length(&g, &res, 1, 1);
  igraph_destroy(&g);
}
```
Here's the report of UBSan:
```
structural_properties.c:281:8: runtime error: division by zero
    #0 0x5253bc in igraph_average_path_length /home/grieve/Research/Testing/igraph/igraph/src/structural_properties.c:281:8
    #1 0x512c93 in main /home/grieve/Research/Testing/igraph/reduced.cpp:9:3
    #2 0x7f9934f3ff69 in __libc_start_main (/usr/lib/libc.so.6+0x20f69)
    #3 0x41a419 in _start (/home/grieve/Research/Testing/igraph/reduced+0x41a419)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior structural_properties.c:281:8
```
Strictly speaking, this is a benign UB, as floating point divide-by-zero is defined to be NaN/Infinity on all platforms that supports IEEE754. On the other hand, I'd say that in this case it's more intuitive to explicitly report an error here rather than silently write NaN/Infinity to `res`.
Also note that UBSan reports several other division-by-zero in `random.c` on my platform. They are mostly trying to obtain IEEE754 NaN/Infinity so they should be benign as well. I would still suggest not doing that though.

**Comment by ntamas** [comment]
For sake of consistency with other igraph functions (e.g., `igraph_density`, which returns NaN if the graph has at most one node, or `igraph_avg_nearest_neighbor_degree`, which returns NaN for isolated vertices), I'd rather stick to returning NaN here as well. (Look for occurrences of `IGRAPH_NAN` in the source code to see other examples). What we can do is to check explicitly whether the denominator in the division is zero and then return `IGRAPH_NAN` explicitly to avoid triggering UBSan.

Can you show me a stacktrace for division by zero errors reported from `random.c`?

**Comment by grievejia** [comment]
Just fgrep `/0.0` in `random.c` and you will see where the issue is. In my testing UBSan fires the division-by-zero warnings in several other places as well, but I wasn't sure earlier if you would consider them as problems that need to be resolved so I did not keep track of them :(
I also found APIs like `igraph_transitivity_undirected()` that takes `igraph_transitivity_mode_t` as a parameter and let the client specify what to expect in the divide-by-zero cases rather than just returning NaN. Maybe the styles of those transitivity functions should also be considered for the sake of consistency?

**Comment by ntamas** [comment]
> Just fgrep /0.0 in random.c and you will see where the issue is.
Strange; this line is in `R_finite()`, which (I think) is only used in the R interface but not the C core itself.

> Maybe the styles of those transitivity functions should also be considered for the sake of consistency?
Maybe in a future version as adding a new argument would break the C API of the library.

**Comment by grievejia** [comment]
> Strange; this line is in `R_finite()`, which (I think) is only used in the R interface but not the C core itself.

Here's a simple C program that triggers it:
```c
#include <igraph.h>
int main() {
  igraph_t g;
  igraph_erdos_renyi_game(&g, IGRAPH_ERDOS_RENYI_GNP, 2, 0.1, IGRAPH_DIRECTED, IGRAPH_NO_LOOPS);
  igraph_destroy(&g);
}
```
Here's the stack trace generated by UBSan:
```
random.c:1559:32: runtime error: division by zero
    #0 0x7d01a4 in R_finite /home/grieve/Research/Testing/igraph/igraph/src/random.c:1559:32
    #1 0x7d0b4b in igraph_rpois /home/grieve/Research/Testing/igraph/igraph/src/random.c:1753:10
    #2 0x7c574c in igraph_rgeom /home/grieve/Research/Testing/igraph/igraph/src/random.c:1934:12
    #3 0x7c54f9 in igraph_rng_get_geom /home/grieve/Research/Testing/igraph/igraph/src/random.c:855:12
    #4 0x5208c6 in igraph_erdos_renyi_game_gnp /home/grieve/Research/Testing/igraph/igraph/src/games.c:610:10
    #5 0x5265e5 in igraph_erdos_renyi_game /home/grieve/Research/Testing/igraph/igraph/src/games.c:804:12
    #6 0x51320e in main /home/grieve/Research/Testing/igraph/reduced.cpp:5:3
    #7 0x7f3b89fd6f69 in __libc_start_main (/usr/lib/libc.so.6+0x20f69)
    #8 0x41a9b9 in _start (/home/grieve/Research/Testing/igraph/reduced+0x41a9b9)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior random.c:1559:32 in 
random.c:1559:50: runtime error: division by zero
    #0 0x7d023a in R_finite /home/grieve/Research/Testing/igraph/igraph/src/random.c:1559:50
    #1 0x7d0b4b in igraph_rpois /home/grieve/Research/Testing/igraph/igraph/src/random.c:1753:10
    #2 0x7c574c in igraph_rgeom /home/grieve/Research/Testing/igraph/igraph/src/random.c:1934:12
    #3 0x7c54f9 in igraph_rng_get_geom /home/grieve/Research/Testing/igraph/igraph/src/random.c:855:12
    #4 0x5208c6 in igraph_erdos_renyi_game_gnp /home/grieve/Research/Testing/igraph/igraph/src/games.c:610:10
    #5 0x5265e5 in igraph_erdos_renyi_game /home/grieve/Research/Testing/igraph/igraph/src/games.c:804:12
    #6 0x51320e in main /home/grieve/Research/Testing/igraph/reduced.cpp:5:3
    #7 0x7f3b89fd6f69 in __libc_start_main (/usr/lib/libc.so.6+0x20f69)
    #8 0x41a9b9 in _start (/home/grieve/Research/Testing/igraph/reduced+0x41a9b9)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior random.c:1559:50 in
```

**Comment by ntamas** [comment]
The commit above fixes the original issue but not the one with `random.c`; that will come later.

**Comment by ntamas** [comment]
It turns out that `R_finite()` should have fallen back to using `finite()` or `isfinite()` when the platform provides it, but there was a bug in `src/random.c` that prevented it from doing so. It should be okay now.

**Comment by ntamas** [comment]
I have started running the igraph test cases with `-fsanitize=address,undefined` to catch more of these, but it turns out that not only igraph but also third-party dependencies that we bundle with igraph (e.g., LAPACK) depend on obtaining NaN and infinity values with division-by-zero. In fact, GCC does not even turn on checking for division-by-zero errors in case of floats with `-fsanitize=undefined`, only Clang seems to do this.

The bottom line is that right now I see no point in getting rid of cases when we use division-by-zero to obtain NaN or infinity because it is not really undefined behaviour in any of the platforms that we are targeting.

**Comment by grievejia** [comment]
I see. Thanks for the efforts!
